### PR TITLE
Web Inspector: Lazily create SourceCodeTextRange for search result objects

### PR DESCRIPTION
**Pull request by patrickangle**
#### 7fb9668bb2d3f662dc4f3be8ad7e062c1ae99ac8
<pre>
Web Inspector: Lazily create SourceCodeTextRange for search result objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=242450">https://bugs.webkit.org/show_bug.cgi?id=242450</a>
rdar://problem/96601780

Reviewed by Devin Rousso.

Improve performance of the Search tab by lazily creating SourceCodeTextRange objects only when needed, since creating
this object is expensive because it tracks changes to the source code, but we never take advantage of that tracking and
just immediately use the info to create the title strings for search results. We also already notify users that they
need to search again if content on the page changes.

This accounts for about 50% of the performance improvements that have taken searching for the letter `a` on webkit.org
from ~13 seconds to under a second (using manual stopwatch testing across 5 tests before and after). The other 50% of
the improvement is from bug 242419 (Web Inspector: Search tab should lazily insert/remove its navigation sidebar tree
items during scrolling). Note that very large sites still pose a problem, but instrumenting search performance has been
tricky because they lock up and never return performance/timing data.

* Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js:
(WI.DOMSearchMatchObject):
(WI.DOMSearchMatchObject.prototype.get textRange):
(WI.DOMSearchMatchObject.prototype.get sourceCodeTextRange):
(WI.DOMSearchMatchObject.prototype.saveIdentityToCookie):
* Source/WebInspectorUI/UserInterface/Models/SourceCodeSearchMatchObject.js:
(WI.SourceCodeSearchMatchObject):
(WI.SourceCodeSearchMatchObject.prototype.get textRange):
(WI.SourceCodeSearchMatchObject.prototype.get sourceCodeTextRange):
(WI.SourceCodeSearchMatchObject.prototype.saveIdentityToCookie):
* Source/WebInspectorUI/UserInterface/Views/SearchResultTreeElement.js:
(WI.SearchResultTreeElement):
(WI.SearchResultTreeElement.truncateAndHighlightTitle):

Canonical link: <a href="https://commits.webkit.org/252277@main">https://commits.webkit.org/252277@main</a>
</pre>

### CODE REVIEW

File: Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js [hunk 1]
```diff
@@ -113,16 +114,21 @@ WI.DOMSearchMatchObject = class DOMSearchMatchObject
...
+
         return this._sourceCodeTextRange;
     }
 
     saveIdentityToCookie(cookie)
     {
         cookie[WI.DOMSearchMatchObject.URLCookieKey] = this._resource.url.hash;
         cookie[WI.DOMSearchMatchObject.TitleKey] = this._title;
-        var textRange = this._sourceCodeTextRange.textRange;
+        var textRange = this.sourceCodeTextRange.textRange;
```
Comment (JosephPecoraro) [inline Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js]: Could this just use `this._textRange`? Why does saving to a cookie need to resolve the sourceCodeTextRange?

File: Source/WebInspectorUI/UserInterface/Models/SourceCodeSearchMatchObject.js [hunk 2]
```diff
@@ -32,15 +32,23 @@ WI.SourceCodeSearchMatchObject = class SourceCodeSearchMatchObject
...
 
     // Public
 
     get sourceCode() { return this._sourceCode; }
     get title() { return this._lineText; }
     get searchTerm() { return this._searchTerm; }
-    get sourceCodeTextRange() { return this._sourceCodeTextRange; }
+    get textRange() { return this._textRange; }
+
+    get sourceCodeTextRange() {
```
Comment (JosephPecoraro) [inline Source/WebInspectorUI/UserInterface/Models/SourceCodeSearchMatchObject.js]: Style: Brace on a newline.

File: Source/WebInspectorUI/UserInterface/Models/SourceCodeSearchMatchObject.js [hunk 3]
```diff
@@ -52,7 +60,7 @@ WI.SourceCodeSearchMatchObject = class SourceCodeSearchMatchObject
         if (this._sourceCode.url)
             cookie[WI.SourceCodeSearchMatchObject.URLCookieKey] = this._sourceCode.url.hash;
 
-        var textRange = this._sourceCodeTextRange.textRange;
+        var textRange = this.sourceCodeTextRange.textRange;
```
Comment (JosephPecoraro) [inline Source/WebInspectorUI/UserInterface/Models/SourceCodeSearchMatchObject.js]: Same, could this be `this._textRange`?

**Comment by JosephPecoraro** [comment]
Nice!!

File: Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js [hunk 4]
```diff
@@ -113,17 +114,21 @@ WI.DOMSearchMatchObject = class DOMSearchMatchObject
         return this._searchTerm;
     }
 
+    get textRange() { return this._textRange; }
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js]: Style: single-line `get` should be moved to the top of the `// Public` section

File: Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js [hunk 5]
```diff
@@ -113,17 +114,21 @@ WI.DOMSearchMatchObject = class DOMSearchMatchObject
...
+
     get sourceCodeTextRange()
     {
+        if (!this._sourceCodeTextRange)
+            this._sourceCodeTextRange = resource.createSourceCodeTextRange(textRange);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js]:
```
this._sourceCodeTextRange ||= this._resource.createSourceCodeTextRange(this._textRange);
```

File: Source/WebInspectorUI/UserInterface/Models/SourceCodeSearchMatchObject.js [hunk 6]
```diff
@@ -32,15 +32,24 @@ WI.SourceCodeSearchMatchObject = class SourceCodeSearchMatchObject
...
     get sourceCode() { return this._sourceCode; }
     get title() { return this._lineText; }
     get searchTerm() { return this._searchTerm; }
-    get sourceCodeTextRange() { return this._sourceCodeTextRange; }
+    get textRange() { return this._textRange; }
+
+    get sourceCodeTextRange()
+    {
+        if (!this._sourceCodeTextRange)
+            this._sourceCodeTextRange = this._sourceCode.createSourceCodeTextRange(this._textRange);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/SourceCodeSearchMatchObject.js]:
```
this._sourceCodeTextRange ||= this._sourceCode.createSourceCodeTextRange(this._textRange);
```

**Review by dcrousso — APPROVED** [review]

File: Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js [hunk 7]
```diff
@@ -115,15 +118,16 @@ WI.DOMSearchMatchObject = class DOMSearchMatchObject
 
     get sourceCodeTextRange()
     {
+        this._sourceCodeTextRange ??= resource.createSourceCodeTextRange(textRange);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js]: s/`resource`/`this._resource`
s/`textRange`/`this._textRange`

File: Source/WebInspectorUI/UserInterface/Models/SourceCodeSearchMatchObject.js [hunk 8]
```diff
@@ -32,15 +32,23 @@ WI.SourceCodeSearchMatchObject = class SourceCodeSearchMatchObject
         this._sourceCode = sourceCode;
         this._lineText = lineText;
         this._searchTerm = searchTerm;
-        this._sourceCodeTextRange = sourceCode.createSourceCodeTextRange(textRange);
+        this._textRange = textRange;
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/SourceCodeSearchMatchObject.js:36]: NIT: should we add a `console.assert(textRange instanceof WI.TextRange, textRange)` above?

File: Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js [hunk 9]
```diff
@@ -34,7 +34,8 @@ WI.DOMSearchMatchObject = class DOMSearchMatchObject
         this._domNode = domNode;
         this._title = title;
         this._searchTerm = searchTerm;
-        this._sourceCodeTextRange = resource.createSourceCodeTextRange(textRange);
+        this._textRange = textRange;
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/DOMSearchMatchObject.js:38]: NIT: should we add a `console.assert(textRange instanceof WI.TextRange, textRange)` above?

**Review by dcrousso — APPROVED** [review]

**Pull request merged**